[UXIT-3480] Refactor upload progress logic and add “Publishing” badge state

barbaraperic opened the pull request:
## Description

- Refactored the upload progress logic to clarify the conceptual hierarchy between **steps**, **stages**, and the overall **upload** process (mainly in the `use-upload-progress` hook and `components`)
- The goal was to make the data model, naming, and types more expressive and consistent across the codebase — improving readability, type safety, and maintainability.

## Key Changes

1. Clarified the data model
- Step → the smallest unit of work (`creating-car`, `announcing-cids`, etc.)

- Stage → a grouping of one or more steps
  - `first-stage` = `creating-car`, `checking-readiness`, `uploading-car`
  - `second-stage` = `announcing-cids`
  - `third-stage` = `finalizing-transaction`

- Upload → the complete process made up of all stages

Defined explicit `StepState`, `StageState`, and `UploadState` types to reflect this structure.

2. Split types into modular files

- Moved upload-related types and util functions into a dedicated `src/types/upload/` and `src/utils/upload/` directory:


3. Renamed ambiguous variables and functions

- `progresses` → `stepStates` to reflect that these hold full step state objects, not just progress values.

- `calculateCompletionStates` → `getUploadOutcome` for clearer intent.

- `getStageBadgeStatus` → `getUploadBadgeStatus` since the badge reflects overall upload state, not an individual stage.


4. Added explicit badge-status logic to reflect `publishing` progress more accurately:

> `In progress` - During:
> 
> - 'creating-car' in progress
> - 'uploading-car' in progress
> - 'checking-readiness' in progress
> - 'announcing-cids' in progress
> - 'finalizing-transaction' in progress
> 
> `Published` - When:
> - 'finalizing-transaction' is completed BUT 'announcing-cids' is not yet completed
> 
> `Pinned` - When:
> - Both 'finalizing-transaction' AND 'announcing-cids' are completed

SgtPooki commented on src/components/upload/progress-card-combined.tsx:
```diff
@@ -1,34 +1,27 @@
-  getCombinedFirstStageProgress: () => number
+  stepStates: Array<StepState>
+  firstStageStatus: StepState['status']
+  firstStageProgress: StepState['progress']
```
I know we already had this code (and I think I added it), but I have not been a fan of explicitly declaring the "firstStage" as it's own prop.. ideally this would be codified in the "StepState" data model and handled appropriately, rather than this explicit naming of the first step.

ultimately though, the goal is simplicity since we want folks to be able to follow the code here easily.

CharlyMartin commented on src/types/upload/step.ts:
```diff
@@ -0,0 +1,17 @@
+
+export interface StepState {
+  step: StepName
+  progress: number // 0–100
```
How about we properly type progress?

```ts
type Progress =
  | 0
  | 1
  | 2
  | 3
  | 4
  | 5
  | 6
  | 7
  | 8
  | 9
  | 10
  | 11
  | 12
  | 13
  | 14
  | 15
  | 16
  | 17
  | 18
  | 19
  | 20
  | 21
  | 22
  | 23
  | 24
  | 25
  | 26
  | 27
  | 28
  | 29
  | 30
  | 31
  | 32
  | 33
  | 34
  | 35
  | 36
  | 37
  | 38
  | 39
  | 40
  | 41
  | 42
  | 43
  | 44
  | 45
  | 46
  | 47
  | 48
  | 49
  | 50
  | 51
  | 52
  | 53
  | 54
  | 55
  | 56
  | 57
  | 58
  | 59
  | 60
  | 61
  | 62
  | 63
  | 64
  | 65
  | 66
  | 67
  | 68
  | 69
  | 70
  | 71
  | 72
  | 73
  | 74
  | 75
  | 76
  | 77
  | 78
  | 79
  | 80
  | 81
  | 82
  | 83
  | 84
  | 85
  | 86
  | 87
  | 88
  | 89
  | 90
  | 91
  | 92
  | 93
  | 94
  | 95
  | 96
  | 97
  | 98
  | 99
  | 100
```

Or perhaps it doesn't to be as granular:

```ts
type Progress = 0 | 10 | 20 | 30 | 40 | 50 | 60 | 70 | 80 | 90 | 100
```

SgtPooki replied:
there are so many cool utility types you can create with some infer and other typescript magic.. https://stackoverflow.com/questions/39494689/is-it-possible-to-restrict-number-to-a-certain-range

but this feels like overkill?

CharlyMartin replied:
Maybe it is yes.

CharlyMartin reviewed (commented):
Super cool Barbara, great one 🎉

barbaraperic commented on src/components/ui/badge-status.tsx:
```diff
@@ -36,6 +38,7 @@ const statusLabels: Record<Status, string | null> = {
   'in-progress': 'In progress',
   completed: 'Complete',
   pinned: 'Pinned',
+  published: 'Published',
```
Sorry, can you clarify for me?

barbaraperic commented:
Thank you both for the comments! I've made additional improvements:

- All `progress` properties are renamed to `stepState` 
- `useUploadProgress` hook now only handles `UploadOutcome` (`hasIpniAnnounceFailure`, `isUploadSuccessful`, and `isUploadFailure`) and `UploadBadgeStatus`  
- `useStepStates` hook provides easy access to individual steps and stage groupings
- `useFirstStageState` calculates the combined progress and status for the first stage
- We were then able to simplify `ProgressCardCombined` to only take `stepStates` as a prop - all calculations are now done with `useFirstStageState` and `useStepStates` within the component
- Added JSDoc comments to explain return values
- Renamed `hasUploadIpniFailure` → `hasIpniAnnounceFailure` (more descriptive)
- Fixed badge status logic to check `announcingCidsStep` completion before showing 'pinned'

@SgtPooki if you agree, I'll tackle "Consider using camelCase for steps instead of kebab-case" in a separate PR 🙏
